File: ios/VocaPhoneApp/App/KeyboardInputLanguages.swift
import UIKit

/// Enabled system keyboards, globe order.
///
/// App-only. `UITextInputMode.activeInputModes` is a required-reason API
/// (`3EC4.1` in the app PrivacyInfo). `LocalModelCatalog` is compiled into the
/// keyboard and Live Activity, which must not ship that call.
enum KeyboardInputLanguages {
    /// UIKit rebuilds `activeInputModes` on every access and its order is not
    /// ours to rely on, while the recommendation list is built from the first
    /// element. Reading it inside `body` let the cards reorder on any redraw —
    /// including the progress tick of a download the user was watching.
    ///
    /// The answer genuinely changes only when the user edits their keyboards,
    /// which they can only do in Settings, so the snapshot is retaken when a
    /// screen opens and when the app comes back.
    @MainActor private static var held: [String]?

    @MainActor
    static var enabledPrimaryLanguages: [String] {
        UITextInputMode.activeInputModes.compactMap(\.primaryLanguage)
    }

    @MainActor
    static var snapshot: [String] {
        if let held { return held }
        let fresh = enabledPrimaryLanguages
        held = fresh
        return fresh
    }

    @MainActor
    static func refresh() {
        held = enabledPrimaryLanguages
    }
}

File: android/app/src/main/java/com/vocahq/vocaphone/ui/KeyboardInputLanguages.kt
package com.vocahq.vocaphone.ui

import android.content.Context
import android.view.inputmethod.InputMethodManager

/**
 * Enabled system keyboards, in the order the user enabled them.
 *
 * App-only. [InputMethodManager.getEnabledInputMethodSubtypeList] is a
 * settings snapshot, not something the keyboard should query on every frame.
 * [com.vocahq.vocaphone.local.LocalModelCatalog] is compiled into the IME,
 * which must not ship this call.
 *
 * The answer genuinely changes only when the user edits their keyboards,
 * which they can only do in Settings, so the snapshot is retaken when a
 * screen opens and when the app comes back.
 */
object KeyboardInputLanguages {
    @Volatile
    private var held: List<String>? = null

    fun enabledPrimaryLanguages(context: Context): List<String> {
        val imm = context.applicationContext
            .getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
            ?: return emptyList()
        val methods = imm.enabledInputMethodList
        val tags = LinkedHashSet<String>()
        for (imi in methods) {
            val subtypes = imm.getEnabledInputMethodSubtypeList(imi, true)
            for (subtype in subtypes) {
                if (subtype.isAuxiliary) continue
                if (subtype.mode == "voice") continue
                @Suppress("DEPRECATION")
                val tag = subtype.languageTag.ifBlank { subtype.locale }
                if (tag.isNotBlank()) tags.add(tag)
            }
        }
        return tags.toList()
    }

    fun snapshot(context: Context): List<String> {
        held?.let { return it }
        val fresh = enabledPrimaryLanguages(context)
        held = fresh
        return fresh
    }

    fun refresh(context: Context) {
        held = enabledPrimaryLanguages(context)
    }
}
